lib/drivers/veres.js

```diff
@@ -768,15 +774,205 @@ api.notes = async options => {
+    rsa: '...'
+  },
+  grantCapability: {
+    ed25519: 'Ed25519SignatureCapabilityAuthorization2018',
+    rsa: '...'
+  },
+  invokeCapability: {
+    ed25519: 'Ed25519SignatureCapabilityAuthorization2018',
+    rsa: '...'
+  }
+};
+
+// get or create params
+function _getParamsForType(didDocument, suiteId, type) {
+  for(const suiteParams of jsonld.getValues(didDocument, suiteId)) {
+    if(suiteParams.type === _suiteParamsTypeMap[suiteId][type]) {
+      return suiteParams;
+    }
+  }
+  // not found, create
+  const params = {
+    type: _suiteParamsTypeMap[suiteId][type],
+    publicKey: []
+  };
+  jsonld.addValue(didDocument, suiteId, params);
+  return params;
+}
+
 api['authn-add'] = async options => {
-  throw new Error('authn-add not implemented');
+  _log(options, 'authn-add', options);
+  const did = options.did;
+  // lock local
+  const release = await _lockDid(did, options);
+  // get local did doc
+  const didDocument = await keyStorage.load(did);
+  // get remote did doc
+  const hostname = _getHostname(options);
+  const ledgerDidDocument = await didcv1.getObservable(
+    {did, hostname, env: options.mode});
+
+  // FIXME: move most of this to ext library
+  const suiteId = _suiteIdMap[options.suite];
+  // get next key not in local or ledger docs
+  let nextKeyId;
+  do {
+    nextKeyId = _nextKeyId(didDocument, suiteId);
+  } while(_hasKeyId(ledgerDidDocument, nextKeyId));
+  const params = _getParamsForType(didDocument, suiteId, options.type);
+  const ledgerParams = _getParamsForType(
+    ledgerDidDocument, suiteId, options.type);
+
+  // public ledger key
+  const newLedgerKey = {
+    id: nextKeyId,
+    owner: didDocument.id
+  };
+  if(options.type === 'ed25519') {
+    newLedgerKey.type = 'Ed25519VerificationKey2018';
+    if(options.public) {
+      newLedgerKey.publicKeyBase58 = options.public;
+    }
+  } else if(options.type === 'rsa') {
+    throw new Error('not implemented');
+  }
+  // clone and add to private local key
+  const newLocalKey = util.deepClone(newLedgerKey);
+  if(options.type === 'ed25519') {
+    if(options.private) {
+      newLocalKey.privateKey = {
+        privateKeyBase58: options.private
+      };
+    }
+  } else if(options.type === 'rsa') {
+    throw new Error('not implemented');
+  }
+  // add to local and remote docs
+  jsonld.addValue(params, 'publicKey', newLedgerKey);
+  jsonld.addValue(ledgerParams, 'publicKey', newLocalKey);
+  //console.log('LEDGER', JSON.stringify(ledgerDidDocument, null, 2));
+  //console.log('LOCAL', JSON.stringify(didDocument, null, 2));
+
+  // FIXME: reusing import code, rename it
+  _import(options, didDocument);
+
+  const operation = didv1.wrap(
+    {didDocument: ledgerDidDocument, operationType: 'update'});
+  //console.log(JSON.stringify(operation, null, 2));
+
+  await release();
+
+  _log(options, 'DID local update successful.');
+
+  //if(options.send) {
+    await _send(options, ledgerDidDocument, didDocument);
```

Comment:
Should be sending `operation` ... or passing `operationType` of `update` into `_send`.